pkg/route/secret/monitor.go

```diff
@@ -0,0 +1,94 @@
+package secret
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+)
+
+type ObjectKey struct {
+	Namespace string
+	Name      string
+}
+
+type singleItemMonitor struct {
+	key         ObjectKey
+	informer    cache.SharedInformer
+	numHandlers atomic.Int32
+	lock        sync.Mutex
+	stopped     bool
+	stopCh      chan struct{}
+}
+
+func NewObjectKey(namespace, name string) ObjectKey {
+	return ObjectKey{
+		Namespace: namespace,
+		Name:      name,
+	}
+}
+
+func newSingleItemMonitor(key ObjectKey, informer cache.SharedInformer) *singleItemMonitor {
+	return &singleItemMonitor{
+		key:      key,
+		informer: informer,
+		stopCh:   make(chan struct{}),
+	}
+}
+
+func (i *singleItemMonitor) HasSynced() bool {
+	return i.informer.HasSynced()
+}
+
+func (i *singleItemMonitor) StartInformer() {
+	klog.Info("starting informer")
+	i.informer.Run(i.stopCh)
+}
+
+func (i *singleItemMonitor) StopInformer() bool {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+
+	if i.stopped {
+		return false
+	}
+	i.stopped = true
+	close(i.stopCh)
+	klog.Info("informer stopped")
+	return true
+}
+
+func (i *singleItemMonitor) AddEventHandler(handler cache.ResourceEventHandler) (SecretEventHandlerRegistration, error) {
```

Comment:
I think our future will be easier to reason about if Add, Remove, and Stop are all locked so they happen in order.

Reply:
Added lock for all : Add, Remove, Stop. Let me know how it looks now. 

I've also added `if i.stopped {}` check for Remove, for handling leakage.